lib/stream/cache.go

```diff
@@ -0,0 +1,198 @@
+//	mu          sync.RWMutex
+//	limit       int
+//	lowest uint64
+//	elems       []*E
+// }
+//
+// func (b *sliceCache[E]) endHeight() uint64 {
+//	b.mu.RLock()
+//	defer b.mu.RUnlock()
+//	if len(b.elems) == 0 {
+//		return 0
+//	}
+//
+//	return b.lowest + uint64(len(b.elems)) - 1
+// }
+//
+// func (b *sliceCache[E]) Set(from uint64, elems []E) {
+//	for i, elem := range elems {
+//		b.setOne(from+uint64(i), elem)
+//	}
+// }
+//
+// func (b *sliceCache[E]) setOne(height uint64, elem E) {
+//	b.mu.Lock()
+//	defer b.mu.Unlock()
+//
+//	defer func() {
+//		// Maybe trim if we grew over limit
+//		if len(b.elems) > b.limit {
+//			b.elems = b.elems[len(b.elems)-b.limit:]
+//		}
+//	}()
+//
+//	start := b.lowest
+//	end := b.endHeight()
+//	if umath.SubtractOrZero(end, height) > uint64(b.limit) {
+//		// Don't add atts that are too far behind, just ignore.
+//		return
+//	}
+//
+//	// If this is the first att, just add it.
+//	if len(b.elems) == 0 {
+//		b.elems = append(b.elems, &elem)
+//		b.lowest = height
+//
+//		return
+//	}
+//	if height < start {
+//		// Add to the front.
+//		growLen := int(start) - int(height)
+//		grow := make([]*E, growLen)
+//		grow[0] = &elem
+//		b.elems = append(grow, b.elems...)
+//		b.lowest = height
+//
+//		return
+//	} else if height > end {
+//		// Add to the back.
+//		growLen := int(height) - int(end)
+//		grow := make([]*E, growLen)
+//		grow[len(grow)-1] = &elem
+//		b.elems = append(b.elems, grow...)
+//
+//		return
+//	}
+//
+//	// Add to middle if not already present
+//	index := int(height) - int(start)
+//	b.elems[index] = &elem
+//
+//	return
+// }
+//
+// func (b *sliceCache[E]) Get(height uint64) []E {
+//	b.mu.RLock()
+//	defer b.mu.RUnlock()
+//
+//	startIndex, ok := umath.Subtract(height, b.lowest)
+//	if !ok {
+//		return nil
+//	}
+//
+//	var resp []E
+//	for i := startIndex; i < uint64(len(b.elems)); i++ {
+//		elem := b.elems[i]
+//		if elem == nil {
+//			break
+//		}
+//
+//		resp = append(resp, *elem)
+//	}
+//
+//	return resp
+// }
+
+// NewMapCache returns a new map based cache using the provided limit.
+func NewMapCache[E any](limit int) Cache[E] {
+	return &mapCache[E]{
+		limit: limit,
+		elems: make(map[uint64]E),
+	}
+}
+
+// mapCache implements Cache storing elements in a map.
+// It will only retain the highest `limit` elements.
+// It assumes elements are roughly sequential and that heights are not sparse.
+type mapCache[E any] struct {
+	limit int // Immutable
+
+	// Mutable fields
+	mu     sync.RWMutex
+	elems  map[uint64]E
+	lowest uint64 // Invalid while elems is empty
+}
+
+// Get returns all strictly sequential elements in the cache from the provided height (inclusive).
+func (c *mapCache[E]) Get(from uint64) ([]E, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	var resp []E
+	for height := from; ; height++ {
+		elem, ok := c.elems[height]
+		if !ok {
+			break
+		} else if height < c.lowest {
+			return nil, errors.New("height < lowest [BUG]") // Sanity check
```

Comment:
why is this a bug? if a cache keeps a limit and the items exceeding the limit gets ejected, then if a relayers want to restart from begining it will request items that have height lower than the lowest height of the current cache state.

Reply:
the contract of `c.lowest` is: "the lower element in the cache". So if you find something lower, then that invariant doesn't hold right?